include/libtorrent/aux_/win_cng.hpp

```diff
@@ -0,0 +1,206 @@
+/*
+
+Copyright (c) 2016, Arvid Norberg
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_WIN_CNG_HPP
+#define TORRENT_WIN_CNG_HPP
+
+#include <vector>
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/aux_/throw.hpp"
+#include "libtorrent/aux_/windows.hpp"
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include <bcrypt.h>
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+namespace libtorrent { namespace aux {
+
+	static void throw_ntstatus_error(char const* name, NTSTATUS status) {
+		char message[1024];
+		std::snprintf(message, sizeof(message),
+			"Error 0x%x returned by %s", status, name);
+		throw_ex<std::runtime_error>(message);
+	}
+
+	enum class cng_algorithm : std::uint8_t {
+		rng = 0,
+		sha1,
+		sha256,
+		sha512
+	};
+
+	inline BCRYPT_ALG_HANDLE cng_open_algorithm_handle(cng_algorithm alg_id)
+	{
+		static LPCWSTR algorithm_str[] =
+		{
+			BCRYPT_RNG_ALGORITHM,
+			BCRYPT_SHA1_ALGORITHM,
+			BCRYPT_SHA256_ALGORITHM,
+			BCRYPT_SHA512_ALGORITHM
+		};
+		LPCWSTR str = algorithm_str[static_cast<uint8_t>(alg_id)];
```

Comment:
the portable spelling is `std::uint8_t`

Reply:
ok, will try to move to LPCWSTR typename template approach.